File: bots-di/server_module/game_state/game_state.py
from typing import Optional

from server_module.game_rules.game_rules import GameRules
from server_module.game_state.armies import Armies
from server_module.game_state.bids import Bids
from server_module.game_state.board_cards import BoardCards
from server_module.game_state.dominance_tokens_usage import DominanceTokensUsage
from server_module.game_state.placed_orders import PlacedOrders
from server_module.game_state.tracks import Tracks
from server_module.game_state.supplies import Supplies
from server_module.game_state.discarded_house_cards import DiscardedHouseCards
from server_module.game_state.power_tokens import PowerTokens
from server_module.game_state.used_mustering_points import UsedMusteringPoints
from server_module.game_state.available_orders import AvailableOrders
from server_module.game_state.combat import Combat


class GameState(dict[str, dict | int]):
    def __init__(
            self,
            armies: Armies,
            placed_orders: PlacedOrders,
            tracks: Tracks,
            supplies: Supplies,
            discarded_house_cards: DiscardedHouseCards,
            power_tokens: PowerTokens,
            dominance_tokens_usage: DominanceTokensUsage,
            used_mustering_points: UsedMusteringPoints,
            available_orders: AvailableOrders,
            bids: Optional[Bids],
            combat: Combat | None,
            wildling_counter: int = 6,
            round_counter: int = 1,
            **kwargs
    ):
        super().__init__(**kwargs)
        self['armies'] = self.armies = armies
        self['placedOrders'] = self.placed_orders = placed_orders
        self['tracks'] = self.tracks = tracks
        self['supplies'] =self.supplies = supplies
        self['discardedHouseCards'] = self.discarded_house_cards = discarded_house_cards
        self['powerTokens'] = self.power_tokens = power_tokens
        self['dominanceTokensUsage'] = self.dominance_tokens_usage = dominance_tokens_usage
        self['usedMusteringPoints'] = self.used_mustering_points = used_mustering_points
        self['availableOrders'] = self.available_orders = available_orders
        self['bids'] = self.bids = bids
        self['combat'] = self.combat = combat
        self['wildlingCounter'] = self.wildling_counter = wildling_counter
        self['roundCounter'] = self.round_counter = round_counter
        self['boardCards'] = self.board_cards = BoardCards()

    @classmethod
    def from_json(cls, json, game_rules: GameRules):
        placed_orders = PlacedOrders(**json['placedOrders']) if 'placedOrders' in json else PlacedOrders()
        available_orders = AvailableOrders.build_from_placed_orders(game_rules, placed_orders)
        return cls(
            Armies(**json['armies']),
            placed_orders,
            Tracks(**json['tracks']),
            Supplies(**json['supplies']),
            DiscardedHouseCards(**json['discardedHouseCards']),
            PowerTokens(**json['powerTokens']),
            DominanceTokensUsage(**json['dominanceTokensUsage']),
            UsedMusteringPoints(**json['usedMusteringPoints']),
            available_orders,
            Bids(**json['bids']) if ('bids' in json and json['bids'] is not None) else None,
            Combat.from_json(json['combat']) if ('combat' in json and json['combat'] is not None) else None,
            json['wildlingCounter'],
            json['roundCounter'],
        )

    def compare(self, other: "GameState") -> bool:
        for key, val in self.items():
            if isinstance(val, dict):
                if not val.compare(other[key]):
                    return False
            if isinstance(val, int):
                if other[key] != val:
                    return False
        return True

File: game-logic-core/scala3MainProject/src/main/scala/fwc/game/GameState.scala
package fwc.game

import enrichment.ExtUPickleHashMap
import fwc.{JsonParsable, JsonSerializable}
import fwc.game.actionPhase.{Combat, DiscardedHouseCards, DominanceTokensUsage}
import fwc.game.eventsPhase.cards.BoardCards
import fwc.game.board.{Armies, TileNumber, Tracks}
import fwc.game.eventsPhase.{Bids, PowerTokens, Supplies, UsedMusteringPoints}
import fwc.game.houses.HouseType
import fwc.game.phases.planningSubPhases.SubPhaseAddOrder
import fwc.game.phases.SubPhase
import fwc.game.planningPhase.{AvailableOrders, Order, OrderType, PlacedOrders}
import ujson.Value

import scala.annotation.tailrec
import scala.util.Try

case class GameState(
                      subPhase: SubPhase,
                      armies: Armies,
                      placedOrders: PlacedOrders,
                      tracks: Tracks,
                      supplies: Supplies,
                      discardedHouseCards: DiscardedHouseCards,
                      powerTokens: PowerTokens,
                    //TODO need to add used board cards
                      boardCards: BoardCards,
                      dominanceTokensUsage: DominanceTokensUsage,
                      usedMusteringPoints: UsedMusteringPoints,
                      availableOrders: AvailableOrders,
                      bids: Bids,
                      combat: Combat = null,
                      wildlingCounter: Int = 6,
                      wildlingsStartedFrom12Points: Option[Boolean] = None,
                      roundCounter: Int = 1,
                      winner: Option[HouseType] = None
                    ) extends JsonSerializable {

  def toPersonalJson(myHouse: HouseType): ujson.Value = {
    val placedOrdersP = getPlacedOrders(myHouse)
    val combatP = getCombat(myHouse)
          
    ujson.Obj(
      toCleanJson.value.addPairs(
        GameStateParts.PlacedOrders.toString -> placedOrdersP.toJson,
        GameStateParts.Combat.toString -> (if combatP == null then ujson.Null else combatP.toJson),
      )
    )
  }


  private def getCombat(myHouse: HouseType) = {
    if combat != null then {
      if combat.attackerHouse == myHouse then
        combat.copy(
          defenderTidesOfBattle = null,
          defenderCard = null
        )
      else if combat.defenderHouse == myHouse then
        combat.copy(
          attackerCard = null,
          attackerTidesOfBattle = null
        )
      else
        combat.copy(
          defenderTidesOfBattle = null,
          defenderCard = null,
          attackerCard = null,
          attackerTidesOfBattle = null
        )
    }
    else null
  }

  private def getPlacedOrders(myHouse: HouseType) = {
    if subPhase.isInstanceOf[SubPhaseAddOrder] then
      placedOrders.copy(placedOrders = placedOrders.placedOrders.map(
        (houseType: HouseType, orders: Map[TileNumber, Order]) =>
          if houseType == myHouse then
            houseType -> orders
          else
            houseType -> orders.map(
              (tileNum: TileNumber, order: Order) =>
                tileNum -> Order(OrderType.ConsolidatePower)
            )
      ))
    else placedOrders
  }

  def toJsonForInputtingPlayer(houses: Seq[HouseType]): ujson.Obj =
    val placedOrdersP =
      if subPhase.isInstanceOf[SubPhaseAddOrder] then
        placedOrders.copy(placedOrders = placedOrders.placedOrders.map(
          (houseType: HouseType, orders: Map[TileNumber, Order]) =>
            if houses.contains(houseType) then
              houseType -> orders
            else
              houseType -> orders.map(
                (tileNum: TileNumber, order: Order) =>
                  tileNum -> Order(OrderType.ConsolidatePower)
              )
        ))
      else placedOrders
    val combatP =
      if combat != null then {
        if houses.contains(combat.attackerHouse) && houses.contains(combat.defenderHouse) then
          combat
        else if houses.contains(combat.defenderHouse) then
          combat.copy(
            attackerCard = null,
            attackerTidesOfBattle = null
          )
        else if houses.contains(combat.attackerHouse) then
          combat.copy(
            defenderCard = null,
            defenderTidesOfBattle = null
          )
        else
          combat.copy(
            defenderTidesOfBattle = null,
            defenderCard = null,
            attackerCard = null,
            attackerTidesOfBattle = null
          )
      }
      else null

    ujson.Obj(
      toCleanJson.value.addPairs(
        GameStateParts.PlacedOrders.toString -> placedOrdersP.toJson,
        GameStateParts.Combat.toString -> (if combatP == null then ujson.Null else combatP.toJson),
      )
    )
  
  def toJson: ujson.Value = {
    ujson.Obj(
      toCleanJson.value.addPairs(
        GameStateParts.BoardCards.toString -> boardCards.toJson,
        GameStateParts.PlacedOrders.toString -> placedOrders.toJson,
        GameStateParts.AvailableOrders.toString -> availableOrders.toJson,
        GameStateParts.Bids.toString -> bids.toJson,
        GameStateParts.Combat.toString -> (if combat == null then ujson.Null else combat.toJson)
      )
    )
  }

  def toCleanJson: ujson.Obj = {
    ujson.Obj(
      GameStateParts.SubPhase.toString -> subPhase.toJson,
      GameStateParts.Armies.toString -> armies.toJson,
      GameStateParts.Tracks.toString -> tracks.toJson,
      GameStateParts.Supplies.toString -> supplies.toJson,
      GameStateParts.DiscardedHouseCards.toString -> discardedHouseCards.toJson,
      GameStateParts.PowerTokens.toString -> powerTokens.toJson,
      GameStateParts.DominanceTokensUsage.toString -> dominanceTokensUsage.toJson,
      GameStateParts.UsedMusteringPoints.toString -> usedMusteringPoints.toJson,
      GameStateParts.WildlingCounter.toString -> wildlingCounter,
      GameStateParts.WildlingsStartedFrom12Points.toString ->
        (if wildlingsStartedFrom12Points.isEmpty then ujson.Null else wildlingsStartedFrom12Points.head),
      GameStateParts.RoundCounter.toString -> roundCounter,
    )
  }

  private def getPart(partName: String, userId: Int, houseType: Option[HouseType] = None): (String, ujson.Value) = {
    partName match
      case GameStateParts.SubPhase.string => GameStateParts.SubPhase.toString -> subPhase.toJson
      case GameStateParts.Armies.string => GameStateParts.Armies.toString -> armies.toJson
      case GameStateParts.Tracks.string => GameStateParts.Tracks.toString -> tracks.toJson
      case GameStateParts.Supplies.string => GameStateParts.Supplies.toString -> supplies.toJson
      case GameStateParts.PowerTokens.string => GameStateParts.PowerTokens.toString -> powerTokens.toJson
      case GameStateParts.DiscardedHouseCards.string => GameStateParts.DiscardedHouseCards.toString -> discardedHouseCards.toJson
      case GameStateParts.UsedMusteringPoints.string => GameStateParts.UsedMusteringPoints.toString -> usedMusteringPoints.toJson
      case GameStateParts.WildlingCounter.string => GameStateParts.WildlingCounter.toString -> wildlingCounter
      case GameStateParts.WildlingsStartedFrom12Points.string => GameStateParts.WildlingsStartedFrom12Points.string
        -> (if wildlingsStartedFrom12Points.isEmpty then ujson.Null else wildlingsStartedFrom12Points.head)
      case GameStateParts.RoundCounter.string => GameStateParts.RoundCounter.toString -> roundCounter
      case GameStateParts.AvailableOrders.string => GameStateParts.AvailableOrders.toString -> availableOrders.toJson
      case GameStateParts.PlacedOrders.string => GameStateParts.PlacedOrders.toString -> (if userId < 0 
        then placedOrders.toJson
        else
          getPlacedOrders(
          if subPhase.isInstanceOf[SubPhaseAddOrder]
          then 
            if houseType.nonEmpty 
            then houseType.head
            else throw new FWCException("No house type provided for placed orders in 'add order' phase.")
          else
            null
          ).toJson)
      case GameStateParts.Combat.string => GameStateParts.Combat.toString -> getCombat(
          if houseType.nonEmpty 
          then houseType.head
          else null
        ).toJson
  }

  def toPartialJson(parts: Seq[String], userId: Int, houseType: Option[HouseType] = None): ujson.Obj = {
    @tailrec
    def buildJson(parts: Seq[String], json: ujson.Obj = ujson.Obj()): ujson.Obj = {
      if parts.isEmpty
      then
        json
      else
        json.obj.addOne(getPart(parts.head, userId,  houseType))
        buildJson(parts.tail, json)
    }
    buildJson(parts)
  }
}

object GameState extends JsonParsable {
  override def fromJson(json: Value): GameState = {
    GameState(
      SubPhase.fromJson(json.obj(GameStateParts.SubPhase.string)),
      Armies.fromJson(json.obj(GameStateParts.Armies.string)),
      PlacedOrders.fromJson(json.obj(GameStateParts.PlacedOrders.string)),
      Tracks.fromJson(json.obj(GameStateParts.Tracks.string)),
      Supplies.fromJson(json.obj(GameStateParts.Supplies.string)),
      DiscardedHouseCards.fromJson(json.obj(GameStateParts.DiscardedHouseCards.string)),
      PowerTokens.fromJson(json.obj(GameStateParts.PowerTokens.string)),
      BoardCards.fromJson(json.obj(GameStateParts.BoardCards.string)),
      DominanceTokensUsage.fromJson(json.obj(GameStateParts.DominanceTokensUsage.string)),
      UsedMusteringPoints.fromJson(json.obj(GameStateParts.UsedMusteringPoints.string)),
      AvailableOrders.fromJson(json.obj(GameStateParts.AvailableOrders.string)),
      Bids.fromJson(json.obj(GameStateParts.Bids.string)),
      Combat.fromJson(json(GameStateParts.Combat.string)),
      json.obj(GameStateParts.WildlingCounter.string).num.toInt,
      Try[Option[Boolean]](json(GameStateParts.WildlingsStartedFrom12Points.string).boolOpt).getOrElse(None),
      json.obj(GameStateParts.RoundCounter.string).num.toInt
    )

  }
}